types/prop-types/index.d.ts

```diff
@@ -1,33 +1,49 @@
+// TypeScript Version: 2.8
 
-export type Validator<T> = (object: T, key: string, componentName: string, ...rest: any[]) => Error | null;
+import { ReactNode, ReactElement } from 'react';
 
-export interface Requireable<T> extends Validator<T> {
-    isRequired: Validator<T>;
+export type IsOptional<T> = undefined | null extends T ? true : undefined extends T ? true : null extends T ? true : false;
+
+export type RequiredKeys<V> = { [K in keyof V]: V[K] extends Validator<infer T> ? IsOptional<T> extends true ? never : K : never }[keyof V];
+export type OptionalKeys<V> = Exclude<keyof V, RequiredKeys<V>>;
+export type InferPropsInner<V> = { [K in keyof V]: InferType<V[K]>; };
+
+export type Validator<T> = (props: object, propName: string, componentName: string, location: string, propFullName: string, _NOMINAL_TYPE_HACK_DO_NOT_USE?: T) => Error | null;
+
+export interface Requireable<T> extends Validator<T | undefined | null> {
+    isRequired: Validator<NonNullable<T>>;
 }
 
-export type ValidationMap<T> = {[K in keyof T]?: Validator<T> };
+export type ValidationMap<T> = { [K in keyof T]-?: Validator<T[K]> };
+// Workaround for type inference, ValidationMap<{}> does not work correctly because of conditional types
+export type AnyValidationMap = { [K in keyof any]: Validator<any> | Requireable<any> };
+
+export type InferType<V> = V extends Validator<infer T> ? T : any;
+export type InferProps<V> =
+    & InferPropsInner<Pick<V, RequiredKeys<V>>>
+    & Partial<InferPropsInner<Pick<V, OptionalKeys<V>>>>;
 
 export const any: Requireable<any>;
-export const array: Requireable<any>;
-export const bool: Requireable<any>;
-export const func: Requireable<any>;
-export const number: Requireable<any>;
-export const object: Requireable<any>;
-export const string: Requireable<any>;
-export const node: Requireable<any>;
-export const element: Requireable<any>;
-export const symbol: Requireable<any>;
-export function instanceOf(expectedClass: {}): Requireable<any>;
-export function oneOf(types: any[]): Requireable<any>;
-export function oneOfType(types: Array<Validator<any>>): Requireable<any>;
-export function arrayOf(type: Validator<any>): Requireable<any>;
-export function objectOf(type: Validator<any>): Requireable<any>;
-export function shape(type: ValidationMap<any>): Requireable<any>;
+export const array: Requireable<any[]>;
+export const bool: Requireable<boolean>;
+export const func: Requireable<(...args: any[]) => any>;
+export const number: Requireable<number>;
+export const object: Requireable<object>;
+export const string: Requireable<string>;
+export const node: Requireable<ReactNode>;
+export const element: Requireable<ReactElement<any>>;
+export const symbol: Requireable<symbol>;
+export function instanceOf<T>(expectedClass: new (...args: any[]) => T): Requireable<T>;
+export function oneOf<T>(types: T[]): Requireable<T>;
+export function oneOfType<T extends Validator<any>>(types: T[]): Requireable<NonNullable<InferType<T>>>;
```

Comment:
As written, I'd think this would find a common subtype of all the input validators, rather than creating a union of tuple members. Maybe not fixable without a 3.0 bump for tuple types, though?

Reply:
Yep, that happened in line 72, when the common type of `ReactNode` and `() => any` is `() => any`. Couldn't get TS to actually infer a tuple type from it without explicitly annotating the type.